homeassistant/components/wemo/__init__.py

```diff
@@ -113,17 +120,14 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     wemo_dispatcher = WemoDispatcher(entry)
     wemo_discovery = WemoDiscovery(hass, wemo_dispatcher, static_conf)
 
-    async def async_stop_wemo(_: Event | None = None) -> None:
-        """Shutdown Wemo subscriptions and subscription thread on exit."""
+    async def async_stop_wemo() -> None:
+        """Shutdown Wemo subscriptions and background threads on exit."""
         _LOGGER.debug("Shutting down WeMo event subscriptions")
         await hass.async_add_executor_job(registry.stop)
         await hass.async_add_executor_job(discovery_responder.stop)
         wemo_discovery.async_stop_discovery()
 
-    entry.async_on_unload(
-        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_stop_wemo)
-    )
-    entry.async_on_unload(async_stop_wemo)
```

Comment:
Why do we now need to call this manually from `async_unload_entry` instead of letting it be called automatically on unload?

Reply:
The ordering matters. Discovery needs to be turned off, so that new devices are not added, before calling `hass.config_entries.async_unload_platforms`. With `async_on_unload` discovery was being turned off after.